src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,689 @@
+        uint256 assetIncrease = totalAssetsBeforeFees - lastTotalAssets;
+
+        // Calculate the performance fee and remove from the total assets before new fees are removed
+        return totalAssetsBeforeFees - ((assetIncrease * fee) / FEE_SCALE);
+    }
+
+    /// @dev Calculate the total assets in the ARM, external withdrawal queue,
+    /// less liquidity assets reserved for the ARM's withdrawal queue and past accrued fees.
+    /// The accrued fees are from the last time fees were calculated.
+    function _rawTotalAssets() internal view returns (uint256) {
+        // Get the assets in the ARM and external withdrawal queue
+        uint256 assets = token0.balanceOf(address(this)) + token1.balanceOf(address(this)) + _externalWithdrawQueue();
+
+        // Load the queue metadata from storage into memory
+        uint256 queuedMem = withdrawsQueued;
+        uint256 claimedMem = withdrawsClaimed;
+
+        // If the ARM becomes insolvent enough that the total value in the ARM and external withdrawal queue
+        // is less than the outstanding withdrawals.
+        if (assets + claimedMem < queuedMem) {
+            return 0;
+        }
+
+        // Need to remove the liquidity assets that have been reserved for the withdrawal queue
+        // and any accrued fees
+        return assets + claimedMem - queuedMem - feesAccrued;
+    }
+
+    /// @dev Hook for calculating the amount of assets in an external withdrawal queue like Lido or OETH
+    /// This is not the ARM's withdrawal queue
+    function _externalWithdrawQueue() internal view virtual returns (uint256 assets);
+
+    /// @notice Calculates the amount of shares for a given amount of liquidity assets
+    function convertToShares(uint256 assets) public view returns (uint256 shares) {
+        uint256 totalAssetsMem = totalAssets();
+        shares = (totalAssetsMem == 0) ? assets : (assets * totalSupply()) / totalAssetsMem;
+    }
+
+    /// @notice Calculates the amount of liquidity assets for a given amount of shares
+    function convertToAssets(uint256 shares) public view returns (uint256 assets) {
+        assets = (shares * totalAssets()) / totalSupply();
+    }
+
+    /// @notice Set the Liquidity Provider Controller contract address.
+    /// Set to a zero address to disable the controller.
+    function setLiquidityProviderController(address _liquidityProviderController) external onlyOwner {
+        liquidityProviderController = _liquidityProviderController;
+
+        emit LiquidityProviderControllerUpdated(_liquidityProviderController);
+    }
+
+    ////////////////////////////////////////////////////
+    ///         Performance Fee Functions
+    ////////////////////////////////////////////////////
+
+    /// @dev Accrues the performance fee based on the increase in total assets
+    /// Needs to be called before any action that changes the liquidity provider shares. eg deposit and redeem
+    function _accruePerformanceFee() internal {
```

Comment:
Hypothetical:

Our prices are (0.98, 0.99). We have only STETH

    Assets: 1000, feesAccrued: 0, Last assets, 1000

All STETH traded for WETH. 

    Assets: 990, feesAccrued: 0, Last assets, 1000

Someone calls deposit or redeem:

    Assets: 990, feesAccrued: 0, Last assets, 990
    
All WETH traded for STETH.

    Assets: 1010, feesAccrued: 0, Last assets, 990
    
Someone calls deposit or redeem:

    Assets: 1010, feesAccrued: 4, Last assets, 1006
    
 The protocol only earned 10 ETH, but we paid out fees on 20ETH.

Reply:
The original idea was the last assets would be a high water mark meaning a fee would not be charged if assets dropped and then returned to the same level. But the implementation does not do that as your example highlights. This was due to complication of deposits and redeems changes the total assets.

Some options:
1. make totalAssets up only after adjusting for deposits and redeems. That is, over report the assets if the available assets drops. 
2. only update last available assets if it goes up after adjusting for deposits or redeems.
3. Make it so stETH can't be sold < 1 and stETH can not be bought > 1. That way each swap is profitable.
4. Calculate the performance fee off the assets per share and only upload the last assets per share if it goes up.

Reply:
I've gone with option 2 for now which is to only update `lastAvailableAssets` if the assets increases and adjusting for the amount of assets on deposits and redeem.
This effectively means `lastAvailableAssets` is a high watermark and the performance fee is only charged when the high water mark is increased. This is adjusted for deposits and redeems at 1:1 which should be ok as the asset is LP WETH.

Reply:
The new method looks good so far. I'm going to mark this as resolved unless I come up with something new.